### PR TITLE
Transaction ordering component

### PR DESCRIPTION
**Pull request by AdoAdoAdo**
## Reasoning behind the pull request
Currently the transactions execution order is re-computed after the actual execution and fed to the outport driver, which serves clients such as indexers or notifiers.
In some cases, e.g multiple smart contract results generated by the same transaction, that need to be executed cross shard in the same destination shard, the execution order for these individual SCRs cannot be correctly determined after execution and all will be treated as a batch and receive the same order.
  
## Proposed changes
Create an ordered collection component that can be used during the Txs execution and collect the transactions and scrs in their correct execution order. 
This PR introduces this component.

## Testing procedure
Created unit tests that fully cover the component.
Regular system test, the component is not integrated with the processors in this PR so there is no relevant scenario that would use the added functionality. Custom scenarios possible after the integration, in a follow up PR.

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: common/ordering/orederedCollection.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+		oc.itemsMap[oc.itemsArray[i]]--
+	}
+}
+
+// GetItems returns the items in the order they were added
+func (oc *orderedCollection) GetItems() []string {
+	oc.mut.RLock()
+	defer oc.mut.RUnlock()
+
+	cpItems := append([]string{}, oc.itemsArray...)
```
Comment (iulianpascalau) [inline common/ordering/orederedCollection.go:79]: I think this is equivalent to:
```GO
cpItems := make([]string, len(oc.itemsArray))
copy(cpItems, oc.itemsArray)
```

File: common/ordering/orederedCollection_test.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	contains = oc.Contains("two")
+	require.True(t, contains)
+
+	contains = oc.Contains("three")
+	require.False(t, contains)
+}
+
+func TestOrderedCollection_IsInterfaceNil(t *testing.T) {
+	oc := ordering.GetNilOrderedCollection()
+	require.True(t, oc.IsInterfaceNil())
```
Comment (iulianpascalau) [inline common/ordering/orederedCollection_test.go:184]: this test can be omitted and instead replace L12 with:
```GO
assert.False(t, check.IfNil(oc))
```
also, a concurrency test on all method would be nice just to test that the mutex locking works

File: common/ordering/orederedCollection.go (hunk 1 above)
Comment (AdoAdoAdo) [reply common/ordering/orederedCollection.go:79]: yes, changed.

**Review by raduchis — DISMISSED** [review]

File: common/ordering/orederedCollection_test.go (hunk 2 above)
Comment (AdoAdoAdo) [reply common/ordering/orederedCollection_test.go]: done

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**